File: 0341-42-3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/0341-42-3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.py
class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:
        def solve(s: str, target_k: int) -> int:
            mp = [0] * 26
            cons = 0
            ans = 0
            l = 0
            n = len(s)
            for r in range(n):
                c = s[r]
                if self.isVowel(c):
                    mp[ord(c) - ord('a')] += 1
                else:
                    cons += 1
                while l <= r and self.check(mp) and cons >= target_k:
                    ans += n - r
                    # Move left pointer
                    left_c = s[l]
                    if self.isVowel(left_c):
                        mp[ord(left_c) - ord('a')] -= 1
                    else:
                        cons -= 1
                    l += 1
            return ans
        
        return solve(word, k) - solve(word, k + 1)
    
    def isVowel(self, c: str) -> bool:
        return c in {'a', 'e', 'i', 'o', 'u'}
    
    def check(self, mp) -> bool:
        return mp[0] > 0 and mp[4] > 0 and mp[8] > 0 and mp[14] > 0 and mp[20] > 0

File: 0341-42-3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/0341-42-3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
class Solution {
    public long countOfSubstrings(String word, int k) {
        return solve(word, k) - solve(word, k + 1);
    }
    
    private long solve(String s, int targetK) {
        int[] mp = new int[26];
        int cons = 0;
        long ans = 0;
        int n = s.length();
        int l = 0;
        for (int r = 0; r < n; r++) {
            char c = s.charAt(r);
            if (isVowel(c)) {
                mp[c - 'a']++;
            } else {
                cons++;
            }
            while (l <= r && check(mp) && cons >= targetK) {
                ans += n - r;
                char leftC = s.charAt(l);
                if (isVowel(leftC)) {
                    mp[leftC - 'a']--;
                } else {
                    cons--;
                }
                l++;
            }
        }
        return ans;
    }
    
    private boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
    
    private boolean check(int[] mp) {
        return mp['a' - 'a'] > 0 && mp['e' - 'a'] > 0 && mp['i' - 'a'] > 0 && mp['o' - 'a'] > 0 && mp['u' - 'a'] > 0;
    }
}